libs/sdk-core/src/reverseswap.rs

```diff
@@ -22,68 +36,438 @@ pub struct CreateReverseSwapResponse {
+    ///
+    /// If `check_if_confirmed` is true, the method checks if the claim tx is confirmed.
+    ///
+    /// If `check_if_confirmed` is false, the method checks if the claim tx is in the mempool or confirmed.
+    async fn is_claim_tx_seen(
+        &self,
+        rsi: &ReverseSwapInfo,
+        check_if_confirmed: bool,
+    ) -> Result<bool> {
+        let lockup_addr = rsi.get_lockup_address(self.config.network)?;
+        let is_claim_tx_seen = self
+            .chain_service
+            .address_transactions(rsi.claim_pubkey.clone())
+            .await?
+            .into_iter()
+            .filter(|t| match check_if_confirmed {
+                true => t.status.block_height.is_some(), // Only consider confirmed txs
+                false => true,                           // Consider all txs
+            })
+            .any(|tx| {
+                tx.vin
+                    .iter()
+                    .any(|vin| vin.prevout.scriptpubkey_address == lockup_addr.to_string())
+            });
+        Ok(is_claim_tx_seen)
+    }
+
+    async fn is_lockup_tx_confirmed(&self, rsi: &ReverseSwapInfo) -> Result<bool> {
+        self.is_lockup_tx_seen(rsi, true).await
+    }
+
+    async fn is_lockup_tx_in_mempool_or_confirmed(&self, rsi: &ReverseSwapInfo) -> Result<bool> {
+        self.is_lockup_tx_seen(rsi, false).await
+    }
+
+    /// The lockup tx is seen when it has an incoming tx of the expected amount
+    ///
+    /// If `check_if_confirmed` is true, the method checks if the lockup tx is confirmed.
+    ///
+    /// If `check_if_confirmed` is false, the method checks if the lockup tx is in the mempool or confirmed.
+    async fn is_lockup_tx_seen(
+        &self,
+        rsi: &ReverseSwapInfo,
+        check_if_confirmed: bool,
+    ) -> Result<bool> {
+        let lockup_addr = rsi.get_lockup_address(self.config.network)?;
+        let is_lockup_tx_seen = self
+            .chain_service
+            .address_transactions(lockup_addr.to_string())
+            .await?
+            .into_iter()
+            .filter(|t| match check_if_confirmed {
+                true => t.status.block_height.is_some(), // Only consider confirmed txs
+                false => true,                           // Consider all txs
+            })
+            .any(|tx| {
+                tx.vin
+                    .iter()
+                    .any(|vin| vin.prevout.value as u64 == rsi.onchain_amount_sat)
+            });
+        Ok(is_lockup_tx_seen)
+    }
+
+    /// Determine the reverse swap status, based on the the states of the lockup and claim tx
+    pub(crate) async fn get_breez_status(
+        &self,
+        rsi: &ReverseSwapInfo,
+    ) -> Result<ReverseSwapStatus> {
+        let payment_hash = rsi.get_preimage_hash().to_hex();
+        let is_invoice_pending = self.persister.get_payment_by_hash(&payment_hash)?.is_some();
+        let status = match is_invoice_pending {
```

Comment:
The `is_invoice_pending` sounds a bit misleading because the `get_payment_by_hash` API returns also completed payments...
I can see cases where we restart the node and the payment is completed but the status is still `Initial` at this point.
In that case we will end in the Cancelled state which is wrong. Also I don't see how we advance the state to `CompletedSeen` or `CompletedConfirmed` since the branch expects false result.

In general once we have a payment (either pending or completed) we should never move to canceled state.

**Suggestion:**

We enter into this method only with the following states (Initial, InProgress, CompletedSeen) and our job is to advance the state if needed.

So for every existing payment (pending or completed) we can do the following based on the status:

--Initial
  |-- if `is_lockup_tx_in_mempool_or_confirmed` transition to InProgress otherwise do nothing.

--InProgress or CompletedSeen
  |----fetch claim tx.
  |------if in mempool -> transition to CompletedSeen
  |------if confirmed -> transition to CompletedConfirmed
  
  If we don't get a result for get_payment_by_hash then it is the only state where we transition to Cancelled.
  
  What do you think @ok300

Reply:
This code (which checks the dynamic / fresh status) is only called in `refresh_reverse_swap()`, which is only called to update the status of monitored swaps in `refresh_monitored_reverse_swaps`.

The monitored swaps are already only of the states you mentioned (Initial, InProgress, CompletedSeen). So this is only called for those states.

Assuming status Initial, a node restart would indeed cancel a rev swap, because it would interrupt GL in the middle of paying it. This is the critical initial phase where the payment either succeeds or fails, anything before that lands "in limbo", e.g. cancelled.

But assuming status InProgress or subsequent, then the pending payment is still found, and the next lookups cover the claim tx.

So maybe I'm missing something, but IMO your suggestion is already implemented. I should restructure the methods for better readability.

Let's have a call tomorrow to clarify.

Reply:
Addressed in 65d07dd1be48609174d2f794fb2c37528b573090